### PR TITLE
BUG: Fix nonrandom first uint32 in MT19937 state initialization.

### LINKED ISSUE #14844: numpy.random.RandomState() fails silently if /dev/urandom returns nothing
For some reason, I get nothing back when I run `cat /dev/urandom | base64 | head -c 5` on my system. This turns numpy deterministic silently.

### Reproducing code example:
Running

```python
import numpy as np
print(np.random.RandomState().get_state()[1][0])
print(np.random.RandomState().get_state()[1][0])
```
should print two different numbers. I get the same number twice.

### Numpy/Python version information:

1.17.3 3.7.5 (default, Oct 17 2019, 12:09:47) 
[GCC 9.2.1 20190827 (Red Hat 9.2.1-1)]

### PR DESCRIPTION
**Pull request by mattip**
Fixes gh-14844. As noted in the comments there by @rkern, the bug was to set the first byte to a value rather than to OR it with that value.

### CODE REVIEW

**Comment by charris** [comment]
I suppose this would be rather tricky to test except by taking a large number of samples.

File: numpy/random/_mt19937.pyx [hunk 1]
```diff
@@ -129,7 +129,7 @@ cdef class MT19937(BitGenerator):
         BitGenerator.__init__(self, seed)
         val = self._seed_seq.generate_state(RK_STATE_LEN, np.uint32)
         # MSB is 1; assuring non-zero initial array
-        self.rng_state.key[0] = 0x80000000UL
+        self.rng_state.key[0] |= 0x80000000UL
         for i in range(1, RK_STATE_LEN):
             self.rng_state.key[i] = val[i]
```
Comment (rkern) [inline numpy/random/_mt19937.pyx]: Check this against the logic in the original code. All of the values must be copied in first, _then_ the MSB gets ORed to 1.

```
for i in range(RK_STATE_LEN):
    self.rng_state.key[i] = val[i]
self.rng_state.key[0] |= 0x80000000UL
```

**Review by rkern — CHANGES_REQUESTED** [review]
The logic is still wrong.

**Comment by rkern** [comment]
For testing, you can give it a `SeedSequence` that generates all 0s.

**Comment by rkern** [comment]
Or rather, a variety of values, in a controlled fashion.

**Comment by rkern** [comment]
```
|15> from numpy.random.bit_generator import ISeedSequence

|16> class FullSeedSequence(ISeedSequence):
...>     def __init__(self, x):
...>         self.x = x
...>     def generate_state(self, n_words, dtype=np.uint32):
...>         return np.full(n_words, self.x, dtype=dtype)
...>     

|17> for i in range(32):
...>     ss = FullSeedSequence(1 << i)
...>     key = MT19937(ss).state['state']['key']
...>     expected = np.full(624, 1 << i, dtype=np.uint32)
...>     expected[0] |= 0x80000000
...>     np.testing.assert_array_equal(key, expected)
...>     
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-17-357821e12562> in <module>()
      4     expected = np.full(624, 1 << i, dtype=np.uint32)
      5     expected[0] |= 0x80000000
----> 6     np.testing.assert_array_equal(key, expected)
      7 

/home/rkern/.edm/envs/python3/lib/python3.6/site-packages/numpy/testing/_private/utils.py in assert_array_equal(x, y, err_msg, verbose)
    916     __tracebackhide__ = True  # Hide traceback for py.test
    917     assert_array_compare(operator.__eq__, x, y, err_msg=err_msg,
--> 918                          verbose=verbose, header='Arrays are not equal')
    919 
    920 

/home/rkern/.edm/envs/python3/lib/python3.6/site-packages/numpy/testing/_private/utils.py in assert_array_compare(comparison, x, y, err_msg, verbose, header, precision, equal_nan, equal_inf)
    839                                 verbose=verbose, header=header,
    840                                 names=('x', 'y'), precision=precision)
--> 841             raise AssertionError(msg)
    842     except ValueError:
    843         import traceback

AssertionError: 
Arrays are not equal

Mismatch: 0.16%
Max absolute difference: 4294967295
Max relative difference: 2.
 x: array([2147483648,          1,          1,          1,          1,
                1,          1,          1,          1,          1,
                1,          1,          1,          1,          1,...
 y: array([2147483649,          1,          1,          1,          1,
                1,          1,          1,          1,          1,
                1,          1,          1,          1,          1,...
```

**Comment by rkern** [comment]
Also, don't forget to fix the other instance of this bug in `_legacy_seeding()`. That will be harder to test, for sure, but hopefully, it will be correct by inspection.

File: numpy/random/_mt19937.pyx (hunk 1 above)
Comment (mattip) [reply numpy/random/_mt19937.pyx]: fixed

**Comment by mattip** [comment]
tests added for both fixes

File: numpy/random/tests/test_randomstate.py [hunk 2]
```diff
@@ -107,6 +107,23 @@ def test_cannot_seed(self):
...
+        for i in range(32):
+            # enter the _legacy_seed(None) path
+            rs.seed()
+            key = rs._bit_generator.state['state']['key']
+            res[i, ...] = key
+        # make sure each the corresponding values at each position in key
+        # are unique. The chances of that failing should be very low.
+        res.shape = (32, -1)
+        for i in range(res.shape[1]):
+            assert len(np.unique(res[:, i])) == 32
```
Comment (rkern) [inline numpy/random/tests/test_randomstate.py]: I might suggest a more targeted regression test for the actual bug: `assert not (res[:, 0] == [0x80000000]*32).all()`

**Review by rkern — CHANGES_REQUESTED** [review]
Looks like there is also this bug in `mt19937_init_by_array`, which is likely the original typo that got translated into the other code paths. It's only used in the legacy seeding for `RandomState`

```
[~]
|4> np.random.RandomState([10, 20, 30]).get_state()[1][0]
2147483648

[~]
|5> np.random.RandomState([10, 20, 31]).get_state()[1][0]
2147483648

[~]
|6> np.random.RandomState([10, 20, 32]).get_state()[1][0]
2147483648

[~]
|8> np.random.MT19937([10, 20, 30]).state['state']['key'][0]
3742924469

[~]
|9> np.random.MT19937([10, 20, 31]).state['state']['key'][0]
2480826684

[~]
|10> np.random.MT19937([10, 20, 32]).state['state']['key'][0]
3174699929
```

File: numpy/random/tests/test_randomstate.py (hunk 2 above)
Comment (mattip) [reply numpy/random/tests/test_randomstate.py]: adopted

**Comment by mattip** [comment]
@rkern tested and fixed for `mt19937_init_by_array`

**Review by rkern — APPROVED** [review]
LGTM! Thanks for seeing this through!

This will definitely require some release notes.

**Comment by charris** [comment]
Was this also a problem with the legacy code in 1.16.x?

**Comment by rkern** [comment]
No, all this was introduced in 1.17

**Comment by charris** [comment]
There is this in 1.16.x `initarray.c`
```
    mt[0] = 0x80000000UL; /* MSB is 1; assuring non-zero initial array */
```

**Comment by rkern** [comment]
Huh. Okay. So I guess we ought to revert that one to maintain our stream-compatibility guarantees.

**Comment by rkern** [comment]
That one dates all the way back. It's still that way in Python. My apologies. I knew that `randomkit`'s `rk_randomseed()` initialization had the `|=`, which meant that `RandomState().get_state()[1][0]` returned something different each time, and we want to do the same whenever a `SeedSequence` is used (regardless of how it was initialized). I didn't thoroughly check all of the code paths in the pre-1.17 code.

**Comment by rkern** [comment]
Do we have legacy-seeding stream tests for the array-seed initialization in addition to the integer-seed route? We probably should.

**Comment by mattip** [comment]
Only `mt19937_init_by_array` should be reverted, the other changes are bugs that made post-randomgen `state['key']` incompatible with pre-randomgen?

**Comment by rkern** [comment]
Yes. Let's add some stream tests to be sure.

**Comment by rkern** [comment]
Hmm, seems like `TestSeed.test_array()` ought to have caught the change.

**Comment by rkern** [comment]
Ah! The lower bits of the first word in the key never actually participate in the algorithm!

https://github.com/numpy/numpy/blob/89f30cbc4c0da2cd88e096d1935f8b9548a8b180/numpy/random/src/mt19937/mt19937.c#L87-L90

`UPPER_MASK = 0x80000000UL` and `LOWER_MASK = 0x7fffffffUL`

`key[0] & UPPER_MASK` gets used and then `key[0]` gets reassigned. `key[0]` gets read later, but only after it gets obliterated.

Nothing was ever broken! It just looked distressing to the user. We can just close this, if we want. Sorry, @mattip!

**Pull request closed without merging**